File: 10.regular-expression-matching/solution.java
class Solution {
    public boolean isMatch(String s, String p) {
        if (p.startsWith("*"))
			return isMatch(s, p.substring(1));
		if (p.isEmpty())
			return s.isEmpty();
		char cur = '\0';
		int j = 0;
        for (int i = 0; i < p.length(); i++)
        {
        	if (p.charAt(i) == '*') {
    			int curJ = j;
        		if (cur == '.')
        		{
        			j = s.length();
        		} else
        		{
        			j--;
        			while (++j < s.length() && s.charAt(j) == cur);
        		}
    			while (true)
    			{
    				if (isMatch(s.substring(j), p.substring(i + 1)))
    					return true;
    				j--;
    				if (j < curJ)
    					return false;
    			}
        	} else {
        		if (cur != '\0')
        		{
                	if (j >= s.length())
                		return false;
                	else if (cur == '.')
	        			j++;
	        		else if (s.charAt(j++) != cur)
	        			return false;
        		}
        		cur = p.charAt(i);
        	}
        }
        char last = p.charAt(p.length() - 1);
        if (last != '*')
        {
        	if (j >= s.length())
        		return false;
        	else if (cur == '.')
    			j++;
    		else if (s.charAt(j++) != cur)
    			return false;
        }
        return j == s.length();
    }
}

File: 10.regular-expression-matching/solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        i = 0
        j = 0
        while True:
            if i == len(p):
                return j == len(s)
            elif i > len(p):
                return False
            if i + 1 < len(p) and p[i + 1] == '*':
                jj = j
                while jj < len(s) and (p[i] == '.' or s[jj] == p[i]):
                    if self.isMatch(s[jj+1:], p[i+2:]):
                        return True
                    jj += 1
                i += 2
            else:
                if j >= len(s):
                    return False
                if p[i] != '.' and p[i] != s[j]:
                    return False
                i += 1
                j += 1
